fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -959,3 +931,188 @@ Future<Standalone<RangeResultRef>> ExclusionInProgressRangeImpl::getRange(ReadYo
                                                                           KeyRangeRef kr) const {
 	return ExclusionInProgressActor(ryw, getKeyRange().begin, kr);
 }
+
+Standalone<RangeResultRef> rywModuleGetRange(ReadYourWritesTransaction* ryw, Standalone<RangeResultRef> res,
+                                             KeyRangeRef kr) {
+	// res is read from database, if ryw enabled, we update it with writes in the transaction
+	Standalone<RangeResultRef> result;
+	RangeMap<Key, std::pair<bool, Optional<Value>>, KeyRangeRef>::Ranges ranges =
+	    ryw->getSpecialKeySpaceWriteMap().containedRanges(kr);
+	RangeMap<Key, std::pair<bool, Optional<Value>>, KeyRangeRef>::iterator iter = ranges.begin();
+	int index = 0;
+	while (iter != ranges.end()) {
+		// add all previous entries into result
+		while (index < res.size() && res[index].key < iter->begin()) {
+			result.push_back(result.arena(), KeyValueRef(res[index].key, res[index].value));
+			result.arena().dependsOn(res.arena());
+			++index;
+		}
+		std::pair<bool, Optional<Value>> entry = iter->value();
+		if (entry.first) {
+			// add the writen entries if exists
+			if (entry.second.present()) {
+				result.push_back_deep(result.arena(), KeyValueRef(iter->begin(), entry.second.get()));
+			}
+			// move index to skip all entries in the iter->range
+			while (index < res.size() && iter->range().contains(res[index].key)) ++index;
+		}
+		++iter;
+	}
+	// add all remaining entries into result
+	while (index < res.size()) {
+		const KeyValueRef& kv = res[index];
+		result.push_back(result.arena(), KeyValueRef(kv.key, kv.value));
+		result.arena().dependsOn(res.arena());
+		++index;
+	}
+	return result;
+}
+
+ACTOR Future<Standalone<RangeResultRef>> getProcessClassActor(ReadYourWritesTransaction* ryw, KeyRef prefix,
+                                                              KeyRangeRef kr) {
+	vector<ProcessData> _workers = wait(getWorkers(&ryw->getTransaction()));
+	auto workers = _workers; // strip const
+	// Note : the sort by string is anti intuition, ex. 1.1.1.1:11 < 1.1.1.1:5
+	std::sort(workers.begin(), workers.end(), [](const ProcessData& lhs, const ProcessData& rhs) {
+		return formatIpPort(lhs.address.ip, lhs.address.port) < formatIpPort(rhs.address.ip, rhs.address.port);
+	});
+	Standalone<RangeResultRef> result;
+	for (auto& w : workers) {
+		// exclude :tls in keys even the network addresss is TLS
+		Key k(prefix.withSuffix(formatIpPort(w.address.ip, w.address.port)));
```

Comment:
Use result.arena() to construct this key

Reply:
Done. Thanks.